transform_constantfold.ml

```diff
@@ -73,17 +132,31 @@ let const_prop ({formals; instrs} : analysis_input) : instructions option =
       if VarMap.equal Approx.equal cur incom then None
       else Some (VarMap.merge merge_approx cur incom)
     in
+
     let update pc cur =
+      (* Before we check if the variable declaration `var x = e` is constant,
+       * we have to normalize `e`. We do this by constant propagating the
+       * current variables, and then constant folding the expression. *)
+      let normalize exp varmap =
+        let try_prop x e =
+          match VarMap.find x varmap with
+          | Approx.Unknown -> e
+          | Approx.Value l -> Edit.replace_var_in_exp x (Constant l) e
+          in
+        fold (VarSet.fold try_prop (expr_vars exp) exp)
+      in
       match[@warning "-4"] instrs.(pc) with
-      | Decl_var (x, Simple (Constant l))
-        ->
-        VarMap.add x (Value l) cur
+      | Decl_var (x, e)
+      | Assign (x, e) ->
+        begin match (normalize e cur) with
```

Comment:
This would be nicer and clearer with an `approx env e` auxiliary function that calls or replaces `normalize` and then converts to `Approx.t`.